Clippit/Word/Assembler/HtmlConverter.cs

```diff
@@ -0,0 +1,395 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Text;
+using System.Text.RegularExpressions;
+using System.Xml;
+using System.Xml.Linq;
+using System.Xml.XPath;
+using DocumentFormat.OpenXml.Packaging;
+using HtmlAgilityPack;
+
+namespace Clippit.Word.Assembler
+{
+    internal static class HtmlConverter
```

Comment:
PowerTools already has an HTML to Wml converter that supports a wide range of HTML tags, css styles, fonts, ol/ui lists, etc.

It seems like it currently builds the whole document, but maybe we can refactor it and reuse it from the document assembler?

@MalcolmJohnston what do you think?

Reply:
@sergey-tihon thanks for coming back to me.

When I first added the support for simple HTML in DocumentAssembler I did consider whether to use the HtmlToWmlConverter.

The reason I didn't go with that approach was partially, because I had a hard time grasping all of the HTML to Wml Converter code, but also because I think the use-cases are different.

As you say currently the Html to Wml Convert is designed for entire documents, and it has a lot of clever functionality for dealing with styles and fonts so that you get a high fidelity conversion.

With the Document Assembler use case I think that we do not need that, and actually it could be retrograde.  The reason is that with Document Assembler we are typically dealing with highly structured documents with specific document styles.  So we wouldn't want the fonts and styles to be altered in the way that Html to Wml Converter does.

If we could find a way to bring HtmlConverter.cs more inline with Html to Wml Converter though then I could see the benefit of that.

Not sure that helps a lot, but some thoughts :)

Reply:
I do not mind keeping two implementations for a while (at least till we refactor the old one and make it universal and usable for both use cases)

However, I do like that we need to add a dependency on `HtmlAgilityPack` just for this simpler implementation. As a nugget package author, we need to depend on as low a version of `HtmlAgilityPack` as possible to minimize the change of version conflict with another package that has a direct or transitive dependency on this package. Same time we need to keep track of all security vulnerabilities inside `HtmlAgilityPack` and bump versions when they are fixed. Can we leave with something like `XElement xhtml` for a while (as it done inside HtmlToWmlConverter)?

Reply:
The reason for using `HtmlAgilityPack `is to handle HTML to XML conversion.  The use-case for me is that we are sometimes pulling through `HTML `that is not well formed or at least it is not `XHTML `so that is the reason for adding `HtmlAgilityPack`.

I understand your concerns about adding dependencies though, I guess the argument can be made that the caller should ensure that the `HTML` is `XHTML` compliant when using this library, and so we only support `XHTML` in Clippit?

That would remove the requirement for `HtmlAgilityPack` being a dependency.  It would also make the conversion of my existing `HtmlConverter `code to something closer to `HtmlToWmlConvertorCore` simpler.

Let me know your thoughts.

Reply:
> I guess the argument can be made that the caller should ensure that the HTML is XHTML compliant when using this library, so we only support XHTML in Clippit?

Yes, exactly, the caller will have a choice: 
1. When `HTML` is known during compile time, it may be easy enough to guarantee that it is `XHTML` compliant and does not pay runtime costs.
2. Add extra dependency to `HtmlAgilityPack` or equivalent, and support non `XHTML` compliant `HTML`.

Reply:
Ok so the action here is to remove HtmlAgilityPack and update the implementation accordingly?

Reply:
yes, thank you

Reply:
Hello Sergey,

Apologies for the delay in getting back to you.  I had another look at this and decided to have a go at refactoring the HtmlConverter code so that it is closer to HtmlToWmlConverterCore.

This is not a complete refactoring as there is a fair amount of work to do, however it does start us along the journey.

What I have done is to create a trimmed down version of the Transform method from HtmlConverterCore in HtmlConverter.  To facilitate this I have made a number of methods and classes `internal` rather than `private`.

The main difference between the two methods are:

1. We do not support all elements in the HtmlConverter version
2. We do not apply paragraph level styles in the HtmlConverter version (because these should be taken from the template)

The implementation as it is adds support for `sub`, `sup` and `s` HTML elements which is an improvement on before.  I also think it provides a pathway towards additional refactoring and further support for HTML tags in HtmlConverter.

I am not intending to do these now, but some of the next steps that would be required are:

1. Decide on a mechansim for not applying Paragraph properties in the Transform method (use existing settings class?)
2. Understand and add support for OL/LI tags (requires some pre-processing based on HtmlToWmlConverterCore
3. Understand how CSS application works and add to HtmlConverter
4. Refactor the Transform method so that it can take a "specification" of the elements that are supported, this could also be considered as an extensibility mechanism i.e. callers could override the default behaviour if needed

This should bring us a lot closer to a common code base for HTML processing which works both for entire documents and for document fragments as required by DocumentAssembler.

Let me know your thoughts.

As an aside did you see that https://github.com/OpenXmlDev/Open-Xml-PowerTools is looking for a new maintainer?  I thought it might be worth approaching them and suggesting they forward people onto Clippit instead?  As a lot of work has already been done in this repository.

Cheers,
Malcolm